x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/ErrorsForCasesWithoutExamplesTestCase.java

```diff
@@ -0,0 +1,167 @@
+
+    protected final List<TestCaseSupplier> paramsToSuppliers(Iterable<Object[]> cases) {
+        List<TestCaseSupplier> result = new ArrayList<>();
+        for (Object[] c : cases) {
+            if (c.length != 1) {
+                throw new IllegalArgumentException("weird layout for test cases");
+            }
+            TestCaseSupplier supplier = (TestCaseSupplier) c[0];
+            result.add(supplier);
+        }
+        return result;
+    }
+
+    public final void test() {
+        int checked = 0;
+        List<TestCaseSupplier> cases = cases();
+        Set<List<DataType>> valid = cases.stream().map(TestCaseSupplier::types).collect(Collectors.toSet());
+        List<Set<DataType>> validPerPosition = AbstractFunctionTestCase.validPerPosition(valid);
+        Iterable<List<DataType>> missingSignatures = missingSignatures(cases, valid)::iterator;
+        for (List<DataType> signature : missingSignatures) {
+            logger.debug("checking {}", signature);
+            List<Expression> args = new ArrayList<>(signature.size());
+            for (DataType type : signature) {
+                args.add(randomLiteral(type));
+            }
+            Expression expression = build(Source.synthetic(sourceForSignature(signature)), args);
+            assertTrue("expected unresolved " + expression, expression.typeResolved().unresolved());
+            assertThat(expression.typeResolved().message(), expectedTypeErrorMatcher(validPerPosition, signature));
+            checked++;
+        }
+        logger.info("checked {} signatures", checked);
+        assertThat("didn't check any signatures", checked, greaterThan(0));
+    }
+
+    private Stream<List<DataType>> missingSignatures(List<TestCaseSupplier> cases, Set<List<DataType>> valid) {
+        return cases.stream()
+            .map(s -> s.types().size())
+            .collect(Collectors.toSet())
+            .stream()
+            .flatMap(AbstractFunctionTestCase::allPermutations)
+            .filter(types -> valid.contains(types) == false)
+            /*
+             * Skip any cases with more than one null. Our tests don't generate
+             * the full combinatorial explosions of all nulls - just a single null.
+             * Hopefully <null>, <null> cases will function the same as <null>, <valid>
+             * cases.
+             */
+            .filter(types -> types.stream().filter(t -> t == DataType.NULL).count() <= 1);
+    }
+
+    protected static String sourceForSignature(List<DataType> signature) {
+        StringBuilder source = new StringBuilder();
+        for (DataType type : signature) {
+            if (false == source.isEmpty()) {
+                source.append(", ");
+            }
+            source.append(type.typeName());
+        }
+        return source.toString();
+    }
+
+    /**
+     * Build the expected error message for an invalid type signature.
+     */
+    protected static String typeErrorMessage(
```

Comment:
For those following along, this is copied out of `AbstactFunctionTestCase` but modified. I copied it because I think that once we apply this pattern globally we'll delete the function I copied it from.